### ISSUE TITLE
When I package my Fyne application, which contains database-related operations, to Android, the application does not open.

### ISSUE THREAD

**Issue by lidldev** [fyne-io/fyne#4514]
### Checklist

- [x] I have searched the [issue tracker](https://github.com/fyne-io/fyne/issues) for open issues that relate to the same problem, before opening a new one.
- [X] This issue only relates to a single bug. I will open new issues for any other problems.

### Describe the bug

I was writing a database-related application, it worked fine when I packaged it without a database, but when I added the database operations, the application did not open and got stuck.

### How to reproduce

I added the sqlite3 package and when I tried to package it to android it gave me an app that wouldn't open

### Screenshots

<img width="217" alt="Screenshot_2024-01-05_at_21 13 43" src="https://github.com/fyne-io/fyne/assets/118509044/ecb4fb36-ad70-454e-92ae-823d30e6ff1a">

### Example code

```
package main

import (
	"database/sql"
	"log"

	"fyne.io/fyne/v2"
	"fyne.io/fyne/v2/app"
	"fyne.io/fyne/v2/container"
	"fyne.io/fyne/v2/layout"
	"fyne.io/fyne/v2/theme"
	"fyne.io/fyne/v2/widget"

	_ "github.com/mattn/go-sqlite3"
)

var db *sql.DB

func openDatabase() (*sql.DB, error) {
	db, err := sql.Open("sqlite3", "kutuphane.db")
	if err != nil {
		return nil, err
	}
	return db, nil
}

func createTable() error {
	_, err := db.Exec(`
		CREATE TABLE IF NOT EXISTS books (
			id INTEGER PRIMARY KEY AUTOINCREMENT,
			book_barcode_number TEXT
			days TEXT
		);
	`)
	return err
}

func insertBook(bookNumber string) error {
	_, err := db.Exec("INSERT INTO books (book_barcode_number) VALUES (?)", bookNumber)
	return err
}

func makeUI(tabs *container.AppTabs) *fyne.Container {
	return container.New(layout.NewGridLayout(1),
		tabs)
}

func makeUI2(entry *widget.Entry, button *widget.Button) *fyne.Container {
	return container.New(layout.NewVBoxLayout(),
		layout.NewSpacer(),
		container.NewVBox(entry),
		container.NewVBox(button),
	)
}

func main() {
	myApp := app.New()
	myWindow := myApp.NewWindow("Kütüphane")
	myWindow.Resize(fyne.NewSize(300, 400))

	var arabicSet []string
	arabicSet = append(arabicSet, "صحيفو دطسيو", "إعدادات", "نشطاء", "يحفظ")

	var err error
	db, err = openDatabase()
	if err != nil {
		log.Fatal(err)
	}
	defer db.Close()

	err = createTable()
	if err != nil {
		log.Fatal(err)
	}

	input := widget.NewEntry()
	input.SetPlaceHolder("Kitap Numarasını Girin")

	tabs := container.NewAppTabs(
		container.NewTabItemWithIcon("Ana Sayfa", theme.HomeIcon(), container.NewVBox(widget.NewButton("Kitap Kaydet", func() {
			w2 := myApp.NewWindow("Kütüphane")
			w2.Resize(fyne.NewSize(300, 400))
			w2.SetContent(makeUI2(input, widget.NewButton("Kaydet", func() {
				bookNumber := input.Text
				if bookNumber == "" {
					fyne.LogError("You Cannot Enter Null Barcode Number", err)
				} else {
					err := insertBook(bookNumber)
					if err != nil {
						log.Fatal(err)
					}
					log.Printf("Book with number %s inserted successfully\n", bookNumber)
				}
			})))
			w2.Show()
			log.Print("Clicked")
		}))),
		container.NewTabItem("Ayarlar", container.NewVBox(widget.NewSelect([]string{"Türkçe", "Arapça"}, func(s string) {

		}))),
	)
	tabs.SetTabLocation(container.TabLocationBottom)

	myWindow.SetContent(makeUI(tabs))

	myWindow.Show()
	myApp.Run()
}
 ```

### Fyne version

2.4.2

### Go compiler version

1.21.5

### Operating system and version

macOS Sonoma 14.2.1

**Comment by Jacalz** [comment]
My guess is that the problem lies with your database dependency (which uses CGo) and not Fyne. Does it support Android?

**Comment by Jacalz** [comment]
I suppose it is using the `os` package internally to try to open a file but the sandbox isn’t letting it. File’s isn’t as much of a thing on mobile as it is on desktop.

**Comment by lidldev** [comment]
> I suppose it is using the `os` package internally to try to open a file but the sandbox isn’t letting it. Files isn’t as much of a thing on mobile as it is on desktop.

So, can't I use database-related things on mobile?

**Comment by andydotxyz** [comment]
> So, can't I use database-related things on mobile?

You can use any library that supports mobile platforms (or are pure Go) and use Reader/Writer to access data instead of assuming a traditional filesystem access.

**Comment by andydotxyz** [comment]
Having said those things, there is no proof that they are the problem with your app.
Typically things that don't work will crash not hang... so something in your app is waiting on startup for something that is not happening on this platform. You could probably confirm that the database is the issue by commenting out the DB init code and trying to run it to see if it works.

**Comment by Jacalz** [comment]
On Android, you can only use regular files if you are within some specific directory that you have access to if I remember correctly. 

I have no idea how your database works and if it works on Android but using a database shouldn't be a problem as long as it works on that os.

I suppose that you might want to try to save your database into `filepath.Join(app.Storage().RootURI(), "katuphane.db")` but I don't have much experience with Android to be honest. I don't think this is a bug in Fyne at least.

**Comment by Jacalz** [comment]
Have you verified that it works without the database?

**Comment by andydotxyz** [comment]
That is an excellent point @Jacalz - the line: `	db, err := sql.Open("sqlite3", "kutuphane.db")
` is almost certainly leading it to try and write to the app directory which is protected and cannot be accessed.

The suggested usage of RootURI is an excellent suggestion. We guarantee that space is writable.

**Comment by lidldev** [comment]
> That is an excellent point @Jacalz - the line: ` db, err := sql.Open("sqlite3", "kutuphane.db") ` is almost certainly leading it to try and write to the app directory which is protected and cannot be accessed.
> 
> The suggested usage of RootURI is an excellent suggestion. We guarantee that space is writable.

Thank you Andy. Is there a video or document explaining how I can use RootURI?

**Comment by Jacalz** [comment]
Here's my code example above but fixed: `filepath.Join(app.Storage().RootURI().Path(), "katuphane.db")`

**Comment by lidldev** [comment]
app.Storage is considered undefined

**Comment by andydotxyz** [comment]
It is a function on the app type. So `a.Storage()` or `myapp.Storage()` depending on what your variable is called.

**Comment by lidldev** [comment]
Oh im sorry 😕 my stupidity

**Comment by lidldev** [comment]
I added the code Block into my main function but it doesnt work app is still stuck at opening screen

**Comment by Jacalz** [comment]
Does it work without the database?

**Comment by lidldev** [comment]
yes

**Comment by lidldev** [comment]
I didn't write the code anyway (database code). I'm not very interested in the database. I wrote it to chat gpt. It seems like there is no problem because it works very well on the computer.

**Comment by Jacalz** [comment]
Just that it works on your computer doesn't mean that it works on Android. Your database package may or may not work there, you'd have to check with them. I'm not sure how much more we can help if the database is what is causing the app to freeze.

**Comment by sdassow** [comment]
A related hint that I picked up from another issue has been to change to the writable directory on startup:
```Go
if err := os.Chdir(myapp.Storage().RootURI().Path()); err != nil {
	log.Fatal(err)
}
```
And then use relative paths (using this myself with a pure Go database though).
